spec/requests/custom_actions_spec.rb

```diff
@@ -220,4 +220,471 @@ def expect_result_to_have_custom_actions_hash
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_security_group_url(nil, @resource))
+    end
+  end
+
+  describe "MiqTemplate" do
+    before(:each) do
+      @resource = FactoryGirl.create(:miq_template)
+      define_custom_button1(@resource)
+    end
+
+    it "queries return custom actions defined" do
+      api_basic_authorize(action_identifier(:templates, :read, :resource_actions, :get))
+
+      get api_template_url(nil, @resource)
+
+      expect(response.parsed_body).to include(
+        "id"      => @resource.id.to_s,
+        "href"    => api_template_url(nil, @resource),
+        "actions" => a_collection_including(a_hash_including("name" => "button1"))
+      )
+    end
+
+    it "accept custom actions" do
+      api_basic_authorize
+
+      post api_template_url(nil, @resource), :params => gen_request(:button1, "key1" => "value1")
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_template_url(nil, @resource))
+    end
+  end
+
+  describe "MiqGroup" do
+    before(:each) do
+      @resource = FactoryGirl.create(:miq_group)
+      define_custom_button1(@resource)
+    end
+
+    it "queries return custom actions defined" do
+      api_basic_authorize(action_identifier(:groups, :read, :resource_actions, :get))
+
+      get api_group_url(nil, @resource)
+
+      expect(response.parsed_body).to include(
+        "id"      => @resource.id.to_s,
+        "href"    => api_group_url(nil, @resource),
+        "actions" => a_collection_including(a_hash_including("name" => "button1"))
+      )
+    end
+
+    it "accept custom actions" do
+      api_basic_authorize
+
+      post api_group_url(nil, @resource), :params => gen_request(:button1, "key1" => "value1")
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_group_url(nil, @resource))
+    end
+  end
+
+  describe "Clusters" do
+    before(:each) do
+      @zone = FactoryGirl.create(:zone, :name => "api_zone")
+      @provider = FactoryGirl.create(:ems_vmware, :zone => @zone)
+      @resource = FactoryGirl.create(:ems_cluster, :ext_management_system => @provider)
+      define_custom_button1(@resource)
+    end
+
+    it "queries return custom actions defined" do
+      api_basic_authorize(action_identifier(:clusters, :read, :resource_actions, :get))
+
+      get api_cluster_url(nil, @resource)
+
+      expect(response.parsed_body).to include(
+        "id"      => @resource.id.to_s,
+        "href"    => api_cluster_url(nil, @resource),
+        "actions" => a_collection_including(a_hash_including("name" => "button1"))
+      )
+    end
+
+    it "accept custom actions" do
+      api_basic_authorize
+
+      post api_cluster_url(nil, @resource), :params => gen_request(:button1, "key1" => "value1")
+
+      expect_single_action_result(:success => true, :message => /.*/, :href => api_cluster_url(nil, @resource))
+    end
+  end
+
+  describe "ExtManagementSystem" do
```

Comment:
would prefer if these reflect the api collection, i.e. Providers,  also if you could sort them in this file, would be easier to look for the related section.  Thanks !!